src/audio/google_rtc_audio_processing.c

```diff
@@ -0,0 +1,385 @@
+	if (!cd)
+		goto fail;
+
+	cd->tuning_handler = comp_data_blob_handler_new(dev);
+	if (!cd->tuning_handler)
+		goto fail;
+
+	cd->state = GoogleRtcAudioProcessingCreate();
+	if (!cd->state) {
+		comp_err(dev, "Failed to initialized GoogleRtcAudioProcessing");
+		goto fail;
+	}
+
+	cd->num_frames = GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE *
+					 GoogleRtcAudioProcessingGetFramesizeInMs(cd->state) / 1000;
+
+	cd->raw_mic_buffer = rballoc(
+		0, SOF_MEM_CAPS_RAM,
+		cd->num_frames * sizeof(cd->raw_mic_buffer[0]));
+	if (!cd->raw_mic_buffer)
+		goto fail;
+	bzero(cd->raw_mic_buffer, cd->num_frames * sizeof(cd->raw_mic_buffer[0]));
+	cd->raw_mic_buffer_index = 0;
+
+	cd->aec_reference_buffer = rballoc(
+		0, SOF_MEM_CAPS_RAM,
+		cd->num_frames * sizeof(cd->aec_reference_buffer[0]));
+	if (!cd->aec_reference_buffer)
+		goto fail;
+	bzero(cd->aec_reference_buffer, cd->num_frames * sizeof(cd->aec_reference_buffer[0]));
+	cd->aec_reference_buffer_index = 0;
+
+	cd->output_buffer = rballoc(
+		0, SOF_MEM_CAPS_RAM,
+		cd->num_frames * sizeof(cd->output_buffer[0]));
+	if (!cd->output_buffer)
+		goto fail;
+	bzero(cd->output_buffer, cd->num_frames * sizeof(cd->output_buffer[0]));
+	cd->output_buffer_index = 0;
+
+	comp_set_drvdata(dev, cd);
+	dev->state = COMP_STATE_READY;
+	comp_dbg(dev, "google_rtc_audio_processing_create(): Ready");
+	return dev;
+fail:
+	comp_err(dev, "google_rtc_audio_processing_create(): Failed");
+	if (cd) {
+		rfree(cd->output_buffer);
+		rfree(cd->aec_reference_buffer);
+		rfree(cd->raw_mic_buffer);
+		comp_data_blob_handler_free(cd->tuning_handler);
+		rfree(cd);
+	}
+	rfree(dev);
+	return NULL;
+}
+
+static void google_rtc_audio_processing_free(struct comp_dev *dev)
+{
+	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "google_rtc_audio_processing_free()");
+
+	GoogleRtcAudioProcessingFree(cd->state);
+	cd->state = NULL;
+	rfree(cd->output_buffer);
+	rfree(cd->aec_reference_buffer);
+	rfree(cd->raw_mic_buffer);
+	comp_data_blob_handler_free(cd->tuning_handler);
+	rfree(cd);
+	rfree(dev);
+}
+
+
+static int google_rtc_audio_processing_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_dbg(dev, "google_rtc_audio_processing_trigger(): %d", cmd);
+
+	return comp_set_state(dev, cmd);
+}
+
+static int google_rtc_audio_processing_prepare(struct comp_dev *dev)
+{
+	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source_buffer;
+	struct list_item *source_buffer_list_item;
+	uint32_t flags;
+
+	comp_dbg(dev, "google_rtc_audio_processing_prepare()");
+
+	/* searching for stream and feedback source buffers */
+	list_for_item(source_buffer_list_item, &dev->bsource_list) {
+		source_buffer = container_of(source_buffer_list_item, struct comp_buffer,
+						 sink_list);
+		buffer_lock(source_buffer, &flags);
+		if (source_buffer->source->comp.type == SOF_COMP_DEMUX)
+			cd->aec_reference = source_buffer;
```

Comment:
IMHO we need a better way to look for a *specific* demux

Reply:
This is just a temporary ad-hoc solution to identify a stream from the other. When doing AEC one need to figure out which stream is the playback reference (aka echo reference) and which one is the microphone. Ideally we should have a way to tag streams in the topology so we can differentiate it not base on where it come from (DMUX/DAI) but base on what they are.

Reply:
@lyakh I have been speaking to @ranj063 about this, the idea is to put "tags" in the buffer so the sink/source can be identified by arbitrary values.